### PR TITLE
Feature custom attributes for users, projects and groups

### PR DESCRIPTION
**Pull request by mipmip**
Hi,

I needed custom_attributes for users and projects. I added custom_attributes for groups to make this PR complete.

Kinds regards,
Pim Snel

### CODE REVIEW

File: lib/gitlab/client/groups.rb [hunk 1]
```diff
@@ -270,5 +270,56 @@ def add_ldap_group_links(id, commonname, group_access, provider)
...
+    # @param  [Integer] group_id The ID of a group.
+    # @return [Gitlab::ObjectifiedHash]
+    def group_custom_attributes(group_id)
+      get("/groups/#{group_id}/custom_attributes")
+    end
+
+    # Gets single group custom_attribute.
+    #
+    # @example
+    #   Gitlab.group_custom_attribute(key, 2)
```
Comment (NARKOZ) [inline lib/gitlab/client/groups.rb:288]:
```suggestion
    #   Gitlab.group_custom_attribute('key', 2)
```

File: lib/gitlab/client/groups.rb [hunk 2]
```diff
@@ -270,5 +270,56 @@ def add_ldap_group_links(id, commonname, group_access, provider)
...
+    def add_group_custom_attribute(key, value, group_id)
+      url = "/groups/#{group_id}/custom_attributes/#{key}"
+      put(url, body: { value: value })
+    end
+
+    # Delete custom_attribute
+    # Will delete a custom_attribute
+    #
+    # @example
+    #   Gitlab.custom_attribute("somekey", 2)
```
Comment (NARKOZ) [inline lib/gitlab/client/groups.rb:315]:
```suggestion
    #   Gitlab.delete_group_custom_attribute('somekey', 2)
```

File: lib/gitlab/client/groups.rb [hunk 3]
```diff
@@ -270,5 +270,56 @@ def add_ldap_group_links(id, commonname, group_access, provider)
...
+    #
+    # @param  [String] key The custom_attribute key to delete
+    # @param  [Integer] group_id The ID of a group.
+    # @return [Boolean]
+    def delete_group_custom_attribute(key, group_id = nil)
+      url = "/groups/#{group_id}/custom_attributes/#{key}"
+      delete(url)
```
Comment (NARKOZ) [inline lib/gitlab/client/groups.rb:322]:
```suggestion
      delete("/groups/#{group_id}/custom_attributes/#{key}")
```

File: lib/gitlab/client/projects.rb [hunk 4]
```diff
@@ -640,5 +640,58 @@ def archive_project(id)
...
+    # Gets project custom_attributes.
+    #
+    # @example
+    #   Gitlab.project_custom_attributes(2)
+    #
+    # @param  [Integer] project_id The ID of a project.
+    # @return [Gitlab::ObjectifiedHash]
+    def project_custom_attributes(project_id)
+      url = "/projects/#{project_id}/custom_attributes"
+      get(url)
```
Comment (NARKOZ) [inline lib/gitlab/client/projects.rb:653]:
```suggestion
      get("/projects/#{project_id}/custom_attributes")
```

File: lib/gitlab/client/projects.rb [hunk 5]
```diff
@@ -640,5 +640,58 @@ def archive_project(id)
...
+    #
+    # @example
+    #   Gitlab.project_custom_attribute(key, 2)
+    #
+    # @param  [String] key The custom_attributes key
+    # @param  [Integer] project_id The ID of a project.
+    # @return [Gitlab::ObjectifiedHash]
+    def project_custom_attribute(key, project_id)
+      url = "/projects/#{project_id}/custom_attributes/#{key}"
+      get(url)
```
Comment (NARKOZ) [inline lib/gitlab/client/projects.rb:666]:
```suggestion
      get("/projects/#{project_id}/custom_attributes/#{key}")
```

File: lib/gitlab/client/projects.rb [hunk 6]
```diff
@@ -640,5 +640,58 @@ def archive_project(id)
...
+    def add_project_custom_attribute(key, value, project_id)
+      url = "/projects/#{project_id}/custom_attributes/#{key}"
+      put(url, body: { value: value })
+    end
+
+    # Delete custom_attribute
+    # Will delete a custom_attribute
+    #
+    # @example
+    #   Gitlab.custom_attribute("somekey", 2)
```
Comment (NARKOZ) [inline lib/gitlab/client/projects.rb:687]:
```suggestion
    #   Gitlab.delete_project_custom_attribute('somekey', 2)
```

File: lib/gitlab/client/projects.rb [hunk 7]
```diff
@@ -640,5 +640,58 @@ def archive_project(id)
...
+    #
+    # @param  [String] key The custom_attribute key to delete
+    # @param  [Integer] project_id The ID of a project.
+    # @return [Boolean]
+    def delete_project_custom_attribute(key, project_id = nil)
+      url = "/projects/#{project_id}/custom_attributes/#{key}"
+      delete(url)
```
Comment (NARKOZ) [inline lib/gitlab/client/projects.rb:694]:
```suggestion
      delete("/projects/#{project_id}/custom_attributes/#{key}")
```

File: lib/gitlab/client/users.rb [hunk 8]
```diff
@@ -279,5 +279,55 @@ def user_search(search, options = {})
...
+    def add_user_custom_attribute(key, value, user_id)
+      url = "/users/#{user_id}/custom_attributes/#{key}"
+      put(url, body: { value: value })
+    end
+
+    # Delete custom_attribute
+    # Will delete a custom_attribute
+    #
+    # @example
+    #   Gitlab.custom_attribute("somekey", 2)
```
Comment (NARKOZ) [inline lib/gitlab/client/users.rb:324]:
```suggestion
    #   Gitlab.delete_user_custom_attribute('somekey', 2)
```

File: lib/gitlab/client/users.rb [hunk 9]
```diff
@@ -279,5 +279,55 @@ def user_search(search, options = {})
...
+    #
+    # @param  [String] key The custom_attribute key to delete
+    # @param  [Integer] user_id The ID of a user.
+    # @return [Boolean]
+    def delete_user_custom_attribute(key, user_id = nil)
```
Comment (NARKOZ) [inline lib/gitlab/client/users.rb:329]: why is a default value for user_id is set to nil?

**Review by NARKOZ — CHANGES_REQUESTED** [review]
Thanks for the PR. I've added comments.

File: lib/gitlab/client/users.rb (hunk 9 above)
Comment (mipmip) [reply lib/gitlab/client/users.rb]: > why is a default value for user_id is set to nil?

Wrong copy of delete_email (Will delete a email an authorized user if no user ID passed). Fixed! 🥂

**Review by NARKOZ — APPROVED** [review]

**Comment by NARKOZ** [comment]
Thank you!

**Pull request merged**